File: Python/Sorting/Quicksort/Quicksort.py
def quickSort(unsorted):
	sort(unsorted,0,len(unsorted)-1)
	
def sort(arr, left, right):
  if left < right:
    partitionValue = partition(arr, left, right)
    sort(arr, left, partitionValue-1)
    sort(arr, partitionValue+1, right)

def partition(arr,left,right):
  leftPointer = left - 1
  pivot = arr[right]
  for i in range(left, right):
    if arr[i] <= pivot:
      leftPointer +=1
      arr[leftPointer], arr[i] = arr[i], arr[leftPointer]
  arr[leftPointer+1], arr[right] = arr[right],arr[leftPointer+1]
  return leftPointer+1
      

	
arr = [5, 7, 4, 2, 9, 66, 22, 6, 11, 33]
print("Original Array")
print(arr)
quickSort(arr)
print("Sorted Array")
print(arr)

File: Java/Sorting/Quicksort/Quicksort.java
public class Quicksort {
	public static void sort(int[] unsorted) {
		sort(unsorted, 0, unsorted.length - 1);
	}
	private static void sort(int[] unsorted, int left, int right) {
		if (right - left <= 0) {
			return;
		}
		int pivot = unsorted[right];
		int partition = partition(unsorted, left, right, pivot);
		sort(unsorted, left, partition - 1);
		sort(unsorted, partition, right);
	}
	private static int partition (int[] arr, int left, int right, int pivot) {
		int leftPointer = left - 1;
		int rightPointer = right;
		while (true) {
			while (arr[++leftPointer] < pivot);
			while (rightPointer > 0 && arr[--rightPointer] > pivot);
			if (leftPointer >= rightPointer) {
				break;
			}
			swap(arr, leftPointer, rightPointer);
		}
		swap(arr, leftPointer, right);
		return leftPointer;
	}
	private static void swap(int[] arr, int one, int two) {
		int temp = arr[one];
		arr[one] = arr[two];
		arr[two] = temp;
	}
}